### PR TITLE
Provide separate canard object and memory pools for each CAN iface

### PR DESCRIPTION
**Pull request by bugobliterator**
Testing:

* Build Size (f103-GPS)
Before
```
Target         Text    Data  BSS    Total 
------------------------------------------
bin/AP_Periph  102181   516  19736  122433
```

After
```
Target         Text    Data  BSS    Total 
------------------------------------------
bin/AP_Periph  104137   520  19732  124389
```

* DNA and messages working on Both buses on CubeOrange-periph in Dual CAN mode. Normal functioning was observed with single Iface on CubeOrange-periph as well

### CODE REVIEW

File: Tools/AP_Periph/can.cpp [hunk 1]
```diff
@@ -113,14 +113,27 @@ HALSITL::CANIface* AP_Periph_FW::can_iface_periph[HAL_NUM_CAN_IFACES];
  * Variables used for dynamic node ID allocation.
  * RTFM at http://uavcan.org/Specification/6._Application_level_functions/#dynamic-node-id-allocation
  */
-static uint32_t send_next_node_id_allocation_request_at_ms; ///< When the next node ID allocation request should be sent
-static uint8_t node_id_allocation_unique_id_offset;         ///< Depends on the stage of the next request
+static uint32_t send_next_node_id_allocation_request_at_ms[HAL_NUM_CAN_IFACES]; ///< When the next node ID allocation request should be sent
+static uint8_t node_id_allocation_unique_id_offset[HAL_NUM_CAN_IFACES];         ///< Depends on the stage of the next request
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:117]: With all these variables with an array of size HAL_NUM_CAN_IFACES... I think it's time to move them into a struct so it's a single variable in an array.

File: Tools/AP_Periph/can.cpp [hunk 2]
```diff
@@ -409,15 +427,16 @@ static void handle_begin_firmware_update(CanardInstance* ins, CanardRxTransfer*
 
 static void handle_allocation_response(CanardInstance* ins, CanardRxTransfer* transfer)
 {
+    uint8_t iface = get_canard_iface_id(ins);
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:430]: const

File: Tools/AP_Periph/can.cpp [hunk 3]
```diff
@@ -977,79 +994,112 @@ static bool shouldAcceptTransfer(const CanardInstance* ins,
     return false;
 }
 
+static void cleanup_stale_trx(uint64_t &timestamp_usec)
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:997]: what is trx? If I'm having to ask, perhaps we need a better name..

File: Tools/AP_Periph/can.cpp [hunk 4]
```diff
@@ -977,79 +994,112 @@ static bool shouldAcceptTransfer(const CanardInstance* ins,
...
+                        &transfer_id[i],
+                        priority,
+                        payload,
+                        payload_len);
+        if (cache == 0) {
+            // we had atleast one interface where transaction was successful
+            success = true;
+        }
+    }
+    return success?0:cache;
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1028]: given two IFACES.. first one is successful, second one fails. The returned result will be the failed one

File: Tools/AP_Periph/can.cpp [hunk 5]
```diff
@@ -977,79 +994,112 @@ static bool shouldAcceptTransfer(const CanardInstance* ins,
...
-        }
-        if (!got_pkt) {
-            break;
+            canardHandleRxFrame(&canard[iface], &rx_frame, timestamp);
         }
     }
 }
 
-static uint16_t pool_peak_percent(void)
+static uint16_t pool_peak_percent(uint8_t iface)
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1100]: const

File: Tools/AP_Periph/can.cpp [hunk 6]
```diff
@@ -1163,31 +1213,96 @@ static void process1HzTasks(uint64_t timestamp_usec)
...
+        return true;
+    }
+
     uint8_t node_id_allocation_transfer_id = 0;
+
+    if (no_iface_finished_dna) {
+        printf("Waiting for dynamic node ID allocation on IF%d... (pool %u)\n", iface, pool_peak_percent(iface));
+    }
+
+    uint32_t now = AP_HAL::native_millis();
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1230]: const

File: Tools/AP_Periph/can.cpp [hunk 7]
```diff
@@ -1163,31 +1213,96 @@ static void process1HzTasks(uint64_t timestamp_usec)
 /*
   wait for dynamic allocation of node ID
  */
-static void can_wait_node_id(void)
+static bool no_iface_finished_dna = true;
+static bool can_do_dna(uint8_t iface)
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1217]: const uint8_t iface

File: Tools/AP_Periph/can.cpp [hunk 8]
```diff
@@ -1096,8 +1144,10 @@ static void process1HzTasks(uint64_t timestamp_usec)
          * The recommended way to establish the minimal size of the memory pool is to stress-test the application and
          * record the worst case memory usage.
          */
-        if (pool_peak_percent() > 70) {
-            printf("WARNING: ENLARGE MEMORY POOL\n");
+        for (uint8_t iface = 0; iface < HAL_NUM_CAN_IFACES; iface++) {
+            if (pool_peak_percent(iface) > 70) {
+                printf("WARNING: ENLARGE MEMORY POOL on Iface %d\n", iface);
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1149]: doesn't this need a (int)iface   ?

File: Tools/AP_Periph/can.cpp [hunk 9]
```diff
@@ -1163,31 +1213,96 @@ static void process1HzTasks(uint64_t timestamp_usec)
...
+
+    // Broadcasting the request
+    const int16_t bcast_res = canardBroadcast(&canard[iface],
+                                                UAVCAN_PROTOCOL_DYNAMIC_NODE_ID_ALLOCATION_SIGNATURE,
+                                                UAVCAN_PROTOCOL_DYNAMIC_NODE_ID_ALLOCATION_ID,
+                                                &node_id_allocation_transfer_id,
+                                                CANARD_TRANSFER_PRIORITY_LOW,
+                                                &allocation_request[0],
+                                                (uint16_t) (uid_size + 1));
+    if (bcast_res < 0)
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1265]: if (...) {
}

File: Tools/AP_Periph/can.cpp [hunk 10]
```diff
@@ -1298,16 +1367,15 @@ void AP_Periph_FW::can_start()
...
 
-    if (PreferredNodeID != CANARD_BROADCAST_NODE_ID) {
-        canardSetLocalNodeID(&canard, PreferredNodeID);
+        if (PreferredNodeID != CANARD_BROADCAST_NODE_ID) {
+            canardSetLocalNodeID(&canard[i], PreferredNodeID);
+        }
     }
 
-    // wait for dynamic node ID allocation
+    // wait for dynamic node ID allocation on atleast one iface
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1378]: "at least" is two words

Comment (magicrub) [reply Tools/AP_Periph/can.cpp:1378]: What is the consequence of an iface running without an allocated Node ID yet?

File: Tools/AP_Periph/can.cpp [hunk 11]
```diff
@@ -1403,6 +1467,10 @@ void AP_Periph_FW::hwesc_telem_update()
 
 void AP_Periph_FW::can_update()
 {
+    for (uint8_t i = 0; i < HAL_NUM_CAN_IFACES; i++) {
+        can_do_dna(i);
```
Comment (magicrub) [inline Tools/AP_Periph/can.cpp:1471]: do we care about the result here?

**Review by magicrub — COMMENTED** [review]
I certainly like the intent of this. I have a few questions and nagging comments but otherwise looks great. I'd like to give this a test for running multiple protocols before it merges

File: Tools/AP_Periph/can.cpp (hunk 4 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:1028]: No, it should be successful, the second one is most likely to fail if it is disconnected.

File: Tools/AP_Periph/can.cpp (hunk 3 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:997]: transactions

File: Tools/AP_Periph/can.cpp (hunk 8 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:1149]: looks like not, otherwise would have failed to build

File: Tools/AP_Periph/can.cpp (hunk 10 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:1378]: we just wait on allocation on that iface, until anything is transmitted. the allocated one goes on as usual

File: Tools/AP_Periph/can.cpp (hunk 11 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:1471]: No, we just attempt at allocation, if it fails, we still move on. If we are here that mean at least one iface got node allocated.

File: Tools/AP_Periph/can.cpp (hunk 1 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:117]: I really don't think it makes a difference in terms of readability. And, you would be increasing size by few bytes for struct padding and program as well 😄 

If we were really about abstraction these variables would be member of AP_Periph class and most of the methods here would be member of the class as well. But that's a discussion for another time.

File: Tools/AP_Periph/can.cpp (hunk 4 above)
Comment (magicrub) [reply Tools/AP_Periph/can.cpp:1028]: So a disconnected iface 2 cache value is what you want to return?

File: Tools/AP_Periph/can.cpp (hunk 3 above)
Comment (magicrub) [reply Tools/AP_Periph/can.cpp]: Rename it to that, please

File: Tools/AP_Periph/can.cpp (hunk 4 above)
Comment (bugobliterator) [reply Tools/AP_Periph/can.cpp:1028]: we return cache on fail, (success == false). if (success == true) its 0, which means successful. If we were to fail we would return the last fail value cached.

Comment (magicrub) [reply Tools/AP_Periph/can.cpp:1028]: Ahh yes. Sorry. I misread it.

File: Tools/AP_Periph/can.cpp (hunk 1 above)
Comment (tridge) [reply Tools/AP_Periph/can.cpp:117]: i'd prefer the structure list, with this type of loop:
 for (auto &s : interfaces) { }

**Review by tridge — CHANGES_REQUESTED** [review]
it would also be great to make the DNA fully async, so we never block for DNA, and just run DNA check for each interface that hasn't had DNA yet. That would be a huge help for the Matek CAN/MSP GPS

**Review by tridge — APPROVED** [review]

**Comment by tridge** [comment]
@bugobliterator nice work! I tested on a f303-MatekGPS and MSP starts up fine without DNA. This means we can fix the hwdef.dat to use DNA

**Pull request merged**